File: src/services/instructions/cmpinv/CMP_W_REGX_REGY.java
package emulator.source.cmpinv

import emulator.engine.CpuContext
 import Instruction from '../Instruction'
import { REGISTER_VALUE_NAME_MAPPER } from '@/constants/registers'

const ASSEMBLER_INSTRUCTION_EXPRESSION = (sdestination, ssource) => `ld.s ${sdestination}, [${ssource}]`

export default class CMP_W_REGX_REGY extends Instruction {
	public CMP_W_REGX_REGY(memory, address, source, 
 destination) {
		super(memory, address, source, destination)
		super.setAssembler("cmp.w " + this.sdest + ", " + this.ssource)
	}

	
	exec ({ context, memory }) {
		int old_a = context[REGISTER_VALUE_NAME_MAPPER[this.destination]] 
		long res = context[REGISTER_VALUE_NAME_MAPPER[this.destination]]  - context[REGISTER_VALUE_NAME_MAPPER[this.source]] 
		markFlags(res, (int)res, context)
		markOverflow(old_a, context[REGISTER_VALUE_NAME_MAPPER[this.source]] , (int)res, context)
		context.pc  += 2
	}
}

File: src/services/instructions/cmpinv/CmpWRegXRegY.js
import Instruction from '../Instruction'
import { REGISTER_VALUE_NAME_MAPPER } from '@/constants/registers'

const ASSEMBLER_INSTRUCTION_EXPRESSION = (sdestination, ssource) => `cmp.w ${sdestination}, ${ssource}`

export default class CmpWRegXRegY extends Instruction {
	constructor (memory, address, source, destination) {
		super(memory, address, source, destination)
		super.setAssembler(ASSEMBLER_INSTRUCTION_EXPRESSION(this.sdestination, this.ssource))
	}

	exec ({ context }) {
		const old_a = context[REGISTER_VALUE_NAME_MAPPER[this.destination]]
		const result = context[REGISTER_VALUE_NAME_MAPPER[this.destination]] - context[REGISTER_VALUE_NAME_MAPPER[this.source]]
		Instruction.markFlags(result, result, context)
		Instruction.markOverflow(old_a, context[REGISTER_VALUE_NAME_MAPPER[this.source]], result, context)
		context.pc  += 2
	}
}
